test/dlc_tests/ec2/test_efa.py

```diff
@@ -0,0 +1,295 @@
+        timeout=300,
+    )
+
+
+@pytest.mark.processor("gpu")
+@pytest.mark.model("N/A")
+@pytest.mark.integration("efa")
+@pytest.mark.usefixtures("sagemaker")
+@pytest.mark.allow_p4de_use
+@pytest.mark.multinode(2)
+@pytest.mark.parametrize("ec2_instance_type,region", EC2_EFA_GPU_INSTANCE_TYPE_AND_REGION)
+@pytest.mark.skipif(
+    is_pr_context() and not is_efa_dedicated(),
+    reason="Skip EFA test in PR context unless explicitly enabled",
+)
+def test_efa_tensorflow(
+    tensorflow_training, efa_ec2_instances, efa_ec2_connections, ec2_instance_type, region, gpu_only
+):
+    number_of_nodes = 2
+    _setup_multinode_efa_instances(
+        tensorflow_training, efa_ec2_instances, efa_ec2_connections, ec2_instance_type, region
+    )
+    master_connection = efa_ec2_connections[0]
+    run_cmd_on_container(MASTER_CONTAINER_NAME, master_connection, EFA_SANITY_TEST_CMD, hide=False)
+    run_cmd_on_container(
+        MASTER_CONTAINER_NAME,
+        master_connection,
+        f"{EFA_INTEGRATION_TEST_CMD} {HOSTS_FILE_LOCATION} {number_of_nodes}",
+        hide=False,
+        timeout=300,
+    )
+
+
+def _setup_multinode_efa_instances(
+    image, efa_ec2_instances, efa_ec2_connections, ec2_instance_type, region
+):
+    """
+    Pull and start test image containers on both master and worker instances, and configure
+    password-less SSH between master and worker nodes.
+    :param image: str DLC image URI to be tested
+    :param efa_ec2_instances: list of tuples of instance_id, keypair_filepath for each instance
+    :param efa_ec2_connections: list of fabric connection objects
+    :param ec2_instance_type: str instance type being used
+    :param region: str region name in which test is being run
+    """
+    # Asynchronously pull docker image on all instances
+    _pull_image_on_all_instances(efa_ec2_connections, image)
+    # Configure master node container
+    master_connection = efa_ec2_connections[0]
+
+    build_all_reduce_perf_promises = []
+    # Run container
+    _setup_container(master_connection, image, MASTER_CONTAINER_NAME)
+    # Build all_reduce_perf binary using nccl-tests
+    promise = run_cmd_on_container(
+        MASTER_CONTAINER_NAME,
+        master_connection,
+        BUILD_ALL_REDUCE_PERF_CMD,
+        timeout=180,
+        asynchronous=True,
+    )
+    build_all_reduce_perf_promises.append(promise)
+    for worker_connection in efa_ec2_connections[1:]:
+        # Run container
+        _setup_container(worker_connection, image, WORKER_CONTAINER_NAME)
+        # Build all_reduce_perf binary using nccl-tests
+        promise = run_cmd_on_container(
+            WORKER_CONTAINER_NAME,
+            worker_connection,
+            BUILD_ALL_REDUCE_PERF_CMD,
+            timeout=180,
+            asynchronous=True,
+        )
+        build_all_reduce_perf_promises.append(promise)
+
+    # Configure master node SSH client-side configurations
+    _setup_master_efa_ssh_config(master_connection)
```

Comment:
nit: efa/mpi setup can be in a bash file, that can directly be executed on the container.

Reply:
Made the decision to keep the SSH/MPI-Hosts setup in the python code so that it can be found and modified more easily outside of the actual test scripts, and to make the test setup more "explainable" and less of a black-box.